File: frontend/src/api/customers.ts
import {
  apiRequest,
} from "./client";


export interface Customer {
  id: string;

  first_name: string;
  last_name?: string | null;

  email?: string | null;
  phone?: string | null;

  nationality?: string | null;
  language?: string | null;

  notes?: string | null;

  created_at?: string | null;
  updated_at?: string | null;
}


export interface CustomerPayload {
  first_name: string;

  last_name?: string | null;

  email?: string | null;
  phone?: string | null;

  nationality?: string | null;
  language?: string | null;

  notes?: string | null;
}


export function getCustomers():
Promise<Customer[]> {
  return apiRequest(
    "/api/customers"
  );
}


export function searchCustomers(
  query: string,
):
Promise<Customer[]> {
  return apiRequest(
    `/api/customers/search?q=${encodeURIComponent(
      query
    )}`
  );
}


export function getCustomer(
  customerId: string,
):
Promise<Customer> {
  return apiRequest(
    `/api/customers/${encodeURIComponent(
      customerId
    )}`
  );
}


export function createCustomer(
  payload: CustomerPayload,
):
Promise<Customer> {
  return apiRequest(
    "/api/customers",
    {
      method: "POST",
      body: payload,
    },
  );
}


export function updateCustomer(
  customerId: string,
  payload:
    Partial<CustomerPayload>,
):
Promise<Customer> {
  return apiRequest(
    `/api/customers/${encodeURIComponent(
      customerId
    )}`,
    {
      method: "PATCH",
      body: payload,
    },
  );
}

File: backend/app/api/routes/customers.py
from fastapi import APIRouter, Depends, Query

from app.api.dependencies import authenticated_user
from app.schemas.customer import (
    CustomerCreate,
    CustomerUpdate,
)
from app.services.customer import CustomerService


router = APIRouter(
    prefix="/api/customers",
    tags=["Customers"],
    dependencies=[
        Depends(authenticated_user),
    ],
)


@router.get("")
async def list_customers(
    limit: int = Query(default=100, ge=1, le=500),
    offset: int = Query(default=0, ge=0),
):
    return CustomerService().list(
        limit=limit,
        offset=offset,
    )


@router.get("/search")
async def search_customers(
    q: str,
):
    return CustomerService().search(q)


@router.get("/{customer_id}")
async def get_customer(customer_id: str):
    customer = CustomerService().get(
        customer_id
    )

    if customer is None:
        raise LookupError("Customer not found")

    return customer


@router.post("")
async def create_customer(
    payload: CustomerCreate,
):
    return CustomerService().create(
        **payload.model_dump()
    )


@router.patch("/{customer_id}")
async def update_customer(
    customer_id: str,
    payload: CustomerUpdate,
):
    return CustomerService().update(
        customer_id,
        payload.model_dump(
            exclude_unset=True
        ),
    )
